frontend/js/src/playlists/Playlist.tsx

```diff
@@ -220,6 +222,25 @@ export default function PlaylistPage() {
     }
     try {
       const jspfTrack = makeJSPFTrack(selectedTrackMetadata);
+
+      // check if the track is already in the playlist
+      const trackMBID = getRecordingMBIDFromJSPFTrack(jspfTrack);
+      const isDuplicate = playlist.track.some((track) => {
+        const existingMBID = getRecordingMBIDFromJSPFTrack(track);
+        return existingMBID === trackMBID;
+      });
+      if (isDuplicate && !dontAskAgain) {
+        const confirmed = await NiceModal.show(DuplicateTrackModal, {
+          message:
+            "This track is already in the playlist. Do you want to add it anyway?",
+          dontAskAgain,
+          setDontAskAgain,
+        });
+        if (!confirmed) {
+          return;
+        }
```

Comment:
Along with my other comment below, I suggest a couple of changes:
1. Get back two values from the modal promise
2. Set an expected return type for the modal
3. Don't pass `setDontAskAgain` down, but instead call it with the new returned boolean value (keeping states more separate between components)
```suggestion
        const [confirmed, dontAskAgainValue] = await NiceModal.show<
          [boolean, boolean],
          any
        >(DuplicateTrackModal, {
          message:
            "This track is already in the playlist. Do you want to add it anyway?",
          dontAskAgain
        });
        setDontAskAgain(dontAskAgainValue);
        if (!confirmed) {
          return;
        }
```